passes/hierarchy/hierarchy.cc

```diff
@@ -710,6 +711,411 @@ RTLIL::Wire *find_implicit_port_wire(Module *module, Cell *cell, const std::stri
+	if (!mods.empty()) {
+		// We can add the binding directly to the module, operating in place.
+		bool did_something = false;
+		log("Applying %s to %d module%s.\n",
+		    desc.c_str(), GetSize(mods), mods.size() > 1 ? "s" : "");
+		for (auto mod : mods)
+			did_something |= binding.bind_into(design, *mod);
+		return did_something;
+	}
+
+	log_error("Couldn't find a target for %s.\n", desc.c_str());
+}
+
+// Worker function for get_cell_type_counts. To avoid double counting, we
+// assume that any module that has already been processed has an entry in
+// counts. Top-level modules get a count of zero added just before calling this
+// function on them.
+static void
+get_cell_type_counts_worker(RTLIL::Design &design,
+                            dict<RTLIL::IdString, int> &counts,
+                            RTLIL::Module &mod)
+{
+	for (auto cell : mod.cells()) {
+		// Set counts[cell->type] to 1 if it isn't defined
+		// yet, or increment if it is.
+		auto pr = counts.insert(std::make_pair(cell->type, 1));
+		if (!pr.second)
+			++pr.first->second;
+
+		// If !pr.second, we have visited this cell type before, so we
+		// don't need to recurse.
+		if (!pr.second)
+			continue;
+
+		RTLIL::Module *child_mod = design.module(cell->type);
+		if (child_mod)
+			get_cell_type_counts_worker(design,
+						    counts, *child_mod);
+	}
+}
+
+// Count the number of times cell type in the design appears as a cell in
+// another module. Cell types for modules that only appear at top-level have a
+// count of zero.
+//
+// Returns a dictionary keyed by module name. Every module will appear in the
+// dictionary, plus any other cell that's used.
+static dict<RTLIL::IdString, int>
+get_cell_type_counts(RTLIL::Design &design)
+{
+	dict<RTLIL::IdString, int> counts;
+	for (auto mod : design.modules()) {
+		counts.insert(std::make_pair(mod->name, 0));
+		get_cell_type_counts_worker(design, counts, *mod);
+	}
+
+	return counts;
+}
+
+// Try to insert cells as instructed by bind directives
+//
+// Returns true if any were inserted. Update *top_mod if necessary.
+static bool insert_bindings(RTLIL::Design &design, RTLIL::Module **top_mod)
+{
+	dict<RTLIL::IdString, int> counts = get_cell_type_counts(design);
+
+	bool bound_something = false;
+
+	// First job: look at any top-level cells that we might need to insert.
+	// These are stored on the Design object itself.
+	for (RTLIL::Binding *binding : design.bindings_)
+		bound_something |=
+			insert_binding(design, counts, nullptr, *binding, top_mod);
+
+	// Now walk the design's modules, looking for binding statements. We
+	// walk all the modules in the design, regardless of whether we have a
+	// top module.
+	//
+	// Note that binding statements aren't allowed to nest (you can't bind
+	// something into a module that you bound in earlier), so we don't have
+	// to do any clever iterative thing.
+	std::vector<RTLIL::Module *> to_visit;
+	std::copy(design.modules().begin(), design.modules().end(),
```

Comment:
Can't we just `to_visit = design.modules();` given the `vector` cast operator of `ObjRange`?

Reply:
Oh, neat! Will change.

Reply:
Ahah! Now I know why I didn't do that! The problem is that we modify design when iterating over the modules. Thinking harder, I'm not sure my original code was quite right: if we bind something into a module that itself specifies a binding, I think we need to make sure we run this on the modified version of the module. The latest force-push should get this right.